File: LeetCode Practice Problems/40. Combination Sum II.cpp
class Solution {
public:
    set<vector<int>> ans;
    void find(vector<int> c,vector<int> p,int target,int index)
    {
        if(target == 0)
        {
            ans.insert(p);
            return;
        }
        for(int i=index+1;i<c.size();i++)
        {
            if(target - c[i] < 0)
                continue;
            p.push_back(c[i]);
            find(c,p,target - c[i],i);
            p.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> p;
        sort(candidates.begin() , candidates.end());
        find(candidates,p,target,-1);
        vector<vector<int>> a;
        for(auto i : ans)
            a.push_back(i);
        return a;
    }
};

File: LeetCode/LeetCode Practice Problems/40-Combination Sum II.java
class Solution {
    
    private void backtrack(List<List<Integer>> ans,List<Integer> res,int[] num,int start,int remain) {
        if(remain < 0) return;
        if(remain == 0) {
            ans.add(new ArrayList<>(res));
            return;
        }
        
        for(int i=start;i<num.length;i++) {
            if(i>start && num[i] == num[i-1]) continue;
            res.add(num[i]);
            backtrack(ans,res,num,i+1,remain-num[i]);
            res.remove(res.size() - 1);
        }
    }
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList();
        Arrays.sort(candidates);
        backtrack(ans,new ArrayList(),candidates,0,target);
        return ans;
    }
}
